### PR TITLE
Topic/t611 032

### PR DESCRIPTION
**Pull request by godunko**

### CODE REVIEW

File: source/text/implementation/vss-strings.adb [hunk 1]
```diff
@@ -186,17 +186,36 @@ package body VSS.Strings is
...
    function Is_Empty (Self : Virtual_String'Class) return Boolean is
+      Handler : constant access Abstract_String_Handler'Class := Self.Handler;
+
    begin
-      return
-        (if Self.Data.In_Place
-         then VSS.Strings.Configuration.In_Place_Handler.Is_Empty (Self.Data)
-         else Self.Data.Handler = null
-           or else Self.Data.Handler.Is_Empty (Self.Data));
+      if Handler /= null then
```
Comment (reznikmm) [inline source/text/implementation/vss-strings.adb:208]: For such statements I prefer to swap then<-> else, so if-condition is "positive":
```ada
if Handler = null then
   return True;
else
   return Handler.Is_Empty (Self.Data);
end if;
```

See [Positive Forms of Logic](https://en.wikibooks.org/wiki/Ada_Style_Guide/Programming_Practices#Positive_Forms_of_Logic) in Ada Style Guide.

File: source/text/implementation/vss-strings.adb [hunk 2]
```diff
@@ -76,16 +76,17 @@ package body VSS.Strings is
...
    function Character_Length
-     (Self : Virtual_String'Class) return Character_Count is
+     (Self : Virtual_String'Class) return Character_Count
+   is
+      Handler : constant access Abstract_String_Handler'Class := Self.Handler;
+
    begin
-      if Self.Data.In_Place then
-         return VSS.Strings.Configuration.In_Place_Handler.Length (Self.Data);
+      if Handler /= null then
```
Comment (reznikmm) [inline source/text/implementation/vss-strings.adb:79]: See Is_Empty

File: source/text/implementation/vss-strings-utf8.adb [hunk 3]
```diff
@@ -70,6 +70,30 @@ package body VSS.Strings.UTF8 is
...
+   ----------------------------
+   -- Before_First_Character --
+   ----------------------------
+
+   overriding procedure Before_First_Character
+     (Self     : UTF8_String_Handler;
+      Data     : String_Data;
+      Position : in out VSS.Strings.Cursor) is
+   begin
+      Position := (Index => 0, UTF8_Offset => 0, UTF16_Offset => 0);
```
Comment (reznikmm) [inline source/text/implementation/vss-strings-utf8.adb:82]: Should it be `UTF8_Offset => VSS.Unicode.UTF8_Code_Unit_Index'Last`?

Comment (godunko) [reply source/text/implementation/vss-strings-utf8.adb:82]: What is rationale behind it?

Comment (reznikmm) [reply source/text/implementation/vss-strings-utf8.adb:82]: to match default initialization.

Comment (godunko) [reply source/text/implementation/vss-strings-utf8.adb:82]: Default initialization should create invalid cursor.

**Pull request merged**